### PR TITLE
AWS Private S2S VPN

### LINKED ISSUE #25530: AWS Site-to-Site VPN Private IP VPNs support
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

> Starting today, AWS Site-to-Site VPN supports the ability to deploy IPSec VPN connections over Direct Connect using private IP addresses.

### New or Affected Resource(s)

* `aws_vpn_connection`

### Potential Terraform Configuration

```hcl
resource "aws_dx_gateway" "test" {
  name                      = "terraform_ipsec_vpn_test"
  amazon_side_asn = "64531"
}

resource "aws_ec2_transit_gateway" "test" {
  amazon_side_asn                  = "64532"
  description                             = "terraform_ipsec_vpn_test"
  transit_gateway_cidr_blocks = [
    "10.0.0.0/24",
  ]
}

resource "aws_customer_gateway" "test" {
  bgp_asn     = 64533
  ip_address = "10.0.0.1"
  type            = "ipsec.1"

  tags = {
    Name = "terraform_ipsec_vpn_test"
  }
}

resource "aws_dx_gateway_association" "test" {
  dx_gateway_id               = aws_dx_gateway.test.id
  associated_gateway_id = aws_ec2_transit_gateway.test.id

  allowed_prefixes = [
    "10.0.0.0/8",
  ]
}

data "aws_ec2_transit_gateway_dx_gateway_attachment" "test" {
  transit_gateway_id = aws_ec2_transit_gateway.test.id
  dx_gateway_id      = aws_dx_gateway.test.id

  depends_on = [
    aws_dx_gateway_association.test
  ]
}

resource "aws_vpn_connection" "test" {
  customer_gateway_id                                 = aws_customer_gateway.test.id
  outside_ip_address_type                            = "PrivateIpv4"
  transit_gateway_id                                       = aws_ec2_transit_gateway.test.id
  transport_transit_gateway_attachment_id = data.aws_ec2_transit_gateway_dx_gateway_attachment.test.id
  type                                                               = "ipsec.1"

  tags = {
    Name = "terraform_ipsec_vpn_test"
  }
}

output "aws_dx_gateway_association_id" {
  value = data.aws_ec2_transit_gateway_dx_gateway_attachment.test.id
}
```

### References

[Announcement](https://aws.amazon.com/about-aws/whats-new/2022/06/aws-site-vpn-introduces-private-ip-security-privacy/).

Requires [AWS SDK for Go v1.44.39](https://github.com/aws/aws-sdk-go/releases/tag/v1.44.39): https://github.com/hashicorp/terraform-provider-aws/pull/25517.

### PR DESCRIPTION
**Pull request by NetDevAutomate**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #25530.

PR for support for Private IP VPNs, a new feature allowing S2S VPN connections to use private ip addresses as the tunnel outside ip address over Direct Connect as transport.

Output from acceptance testing:

```
$  make testacc TESTS=TestAccSiteVPNConnection_ PKG=ec2
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/ec2/... -v -count 1 -parallel 20 -run='TestAccSiteVPNConnection_'  -timeout 180m
=== RUN   TestAccSiteVPNConnection_basic
=== PAUSE TestAccSiteVPNConnection_basic
=== RUN   TestAccSiteVPNConnection_transitGatewayID
=== PAUSE TestAccSiteVPNConnection_transitGatewayID
=== RUN   TestAccSiteVPNConnection_tunnel1InsideCIDR
=== PAUSE TestAccSiteVPNConnection_tunnel1InsideCIDR
=== RUN   TestAccSiteVPNConnection_tunnel1InsideIPv6CIDR
=== PAUSE TestAccSiteVPNConnection_tunnel1InsideIPv6CIDR
=== RUN   TestAccSiteVPNConnection_tunnel1PreSharedKey
=== PAUSE TestAccSiteVPNConnection_tunnel1PreSharedKey
=== RUN   TestAccSiteVPNConnection_tunnelOptions
=== PAUSE TestAccSiteVPNConnection_tunnelOptions
=== RUN   TestAccSiteVPNConnection_tunnelOptionsLesser
=== PAUSE TestAccSiteVPNConnection_tunnelOptionsLesser
=== RUN   TestAccSiteVPNConnection_staticRoutes
=== PAUSE TestAccSiteVPNConnection_staticRoutes
=== RUN   TestAccSiteVPNConnection_outsideAddressTypePrivate
=== PAUSE TestAccSiteVPNConnection_outsideAddressTypePrivate
=== RUN   TestAccSiteVPNConnection_outsideAddressTypePublic
=== PAUSE TestAccSiteVPNConnection_outsideAddressTypePublic
=== RUN   TestAccSiteVPNConnection_enableAcceleration
=== PAUSE TestAccSiteVPNConnection_enableAcceleration
=== RUN   TestAccSiteVPNConnection_ipv6
=== PAUSE TestAccSiteVPNConnection_ipv6
=== RUN   TestAccSiteVPNConnection_tags
=== PAUSE TestAccSiteVPNConnection_tags
=== RUN   TestAccSiteVPNConnection_specifyIPv4
=== PAUSE TestAccSiteVPNConnection_specifyIPv4
=== RUN   TestAccSiteVPNConnection_specifyIPv6
=== PAUSE TestAccSiteVPNConnection_specifyIPv6
=== RUN   TestAccSiteVPNConnection_disappears
=== PAUSE TestAccSiteVPNConnection_disappears
=== RUN   TestAccSiteVPNConnection_updateCustomerGatewayID
=== PAUSE TestAccSiteVPNConnection_updateCustomerGatewayID
=== RUN   TestAccSiteVPNConnection_updateVPNGatewayID
=== PAUSE TestAccSiteVPNConnection_updateVPNGatewayID
=== RUN   TestAccSiteVPNConnection_updateTransitGatewayID
=== PAUSE TestAccSiteVPNConnection_updateTransitGatewayID
=== RUN   TestAccSiteVPNConnection_vpnGatewayIDToTransitGatewayID
=== PAUSE TestAccSiteVPNConnection_vpnGatewayIDToTransitGatewayID
=== RUN   TestAccSiteVPNConnection_transitGatewayIDToVPNGatewayID
=== PAUSE TestAccSiteVPNConnection_transitGatewayIDToVPNGatewayID
=== CONT  TestAccSiteVPNConnection_basic
=== CONT  TestAccSiteVPNConnection_ipv6
=== CONT  TestAccSiteVPNConnection_updateCustomerGatewayID
=== CONT  TestAccSiteVPNConnection_specifyIPv6
=== CONT  TestAccSiteVPNConnection_disappears
=== CONT  TestAccSiteVPNConnection_vpnGatewayIDToTransitGatewayID
=== CONT  TestAccSiteVPNConnection_tunnelOptionsLesser
=== CONT  TestAccSiteVPNConnection_specifyIPv4
=== CONT  TestAccSiteVPNConnection_outsideAddressTypePrivate
=== CONT  TestAccSiteVPNConnection_enableAcceleration
=== CONT  TestAccSiteVPNConnection_tunnel1InsideIPv6CIDR
=== CONT  TestAccSiteVPNConnection_tunnel1InsideCIDR
=== CONT  TestAccSiteVPNConnection_tunnel1PreSharedKey
=== CONT  TestAccSiteVPNConnection_staticRoutes
=== CONT  TestAccSiteVPNConnection_tags
=== CONT  TestAccSiteVPNConnection_updateVPNGatewayID
=== CONT  TestAccSiteVPNConnection_outsideAddressTypePublic
=== CONT  TestAccSiteVPNConnection_updateTransitGatewayID
=== CONT  TestAccSiteVPNConnection_tunnelOptions
=== CONT  TestAccSiteVPNConnection_transitGatewayIDToVPNGatewayID
--- PASS: TestAccSiteVPNConnection_disappears (237.93s)
=== CONT  TestAccSiteVPNConnection_transitGatewayID
--- PASS: TestAccSiteVPNConnection_tunnel1InsideCIDR (240.27s)
--- PASS: TestAccSiteVPNConnection_tunnel1PreSharedKey (240.32s)
--- PASS: TestAccSiteVPNConnection_staticRoutes (243.16s)
--- PASS: TestAccSiteVPNConnection_outsideAddressTypePublic (243.22s)
--- PASS: TestAccSiteVPNConnection_tunnelOptions (243.24s)
--- PASS: TestAccSiteVPNConnection_basic (245.86s)
--- PASS: TestAccSiteVPNConnection_tags (268.53s)
--- PASS: TestAccSiteVPNConnection_updateCustomerGatewayID (465.99s)
--- PASS: TestAccSiteVPNConnection_updateVPNGatewayID (499.75s)
--- PASS: TestAccSiteVPNConnection_specifyIPv4 (561.93s)
--- PASS: TestAccSiteVPNConnection_transitGatewayID (791.96s)
--- PASS: TestAccSiteVPNConnection_updateTransitGatewayID (1148.08s)
=== CONT  TestAccSiteVPNConnection_tunnelOptionsLesser
    vpnsite_connection_test.go:540: Step 5/5 error: Check failed: 3 errors occurred:
        	* Check 12/52 error: aws_vpn_connection.test: list or set attribute 'tunnel1_phase1_dh_group_numbers' must be checked by element count key (tunnel1_phase1_dh_group_numbers.#) or element value keys (e.g. tunnel1_phase1_dh_group_numbers.0). Set element value checks should use TestCheckTypeSet functions instead.
        	* Check 13/52 error: aws_vpn_connection.test: list or set attribute 'tunnel1_phase1_encryption_algorithms' must be checked by element count key (tunnel1_phase1_encryption_algorithms.#) or element value keys (e.g. tunnel1_phase1_encryption_algorithms.0). Set element value checks should use TestCheckTypeSet functions instead.
        	* Check 14/52 error: aws_vpn_connection.test: list or set attribute 'tunnel1_phase1_integrity_algorithms' must be checked by element count key (tunnel1_phase1_integrity_algorithms.#) or element value keys (e.g. tunnel1_phase1_integrity_algorithms.0). Set element value checks should use TestCheckTypeSet functions instead.

--- FAIL: TestAccSiteVPNConnection_tunnelOptionsLesser (1182.01s)
--- PASS: TestAccSiteVPNConnection_tunnel1InsideIPv6CIDR (1238.79s)
--- PASS: TestAccSiteVPNConnection_transitGatewayIDToVPNGatewayID (1240.33s)
--- PASS: TestAccSiteVPNConnection_outsideAddressTypePrivate (1454.86s)
--- PASS: TestAccSiteVPNConnection_specifyIPv6 (1653.20s)
--- PASS: TestAccSiteVPNConnection_enableAcceleration (1693.20s)
--- PASS: TestAccSiteVPNConnection_vpnGatewayIDToTransitGatewayID (1694.92s)
--- PASS: TestAccSiteVPNConnection_ipv6 (1773.72s)
FAIL
FAIL	github.com/hashicorp/terraform-provider-aws/internal/service/ec2	1776.162s

** Note **
Acceptance tests TestAccSiteVPNConnection_tunnelOptionsLesser fail, however these also failed on previous versions

...
```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -89,3 +89,5 @@ require (
 	google.golang.org/grpc v1.46.0 // indirect
 	google.golang.org/protobuf v1.28.0 // indirect
 )
+
+replace github.com/aws/aws-sdk-go => /Users/taylaand/nda/apollo/env/AWSGoSDK-Release/var/tmp/release-automation/staging-sdk1201368572/sdk/src/github.com/aws/aws-sdk-go
```
Comment (AdamTylerLynch) [inline go.mod:93]: Please remove the go.mod changes

File: go.sum [hunk 2]
```diff
@@ -24,10 +24,6 @@ github.com/apparentlymart/go-textseg/v12 v12.0.0/go.mod h1:S/4uRK2UtaQttw1GenVJE
 github.com/apparentlymart/go-textseg/v13 v13.0.0 h1:Y+KvPE1NYz0xl601PVImeQfFyEy6iT90AvPUL1NNfNw=
 github.com/apparentlymart/go-textseg/v13 v13.0.0/go.mod h1:ZK2fH7c4NqDTLtiYLvIkEghdlcqw7yxLeM89kiTRPUo=
 github.com/armon/go-socks5 v0.0.0-20160902184237-e75332964ef5/go.mod h1:wHh0iHkYZB8zMSxRWpUBQtwG5a7fFgvEO+odwuTv2gs=
-github.com/aws/aws-sdk-go v1.42.18/go.mod h1:585smgzpB/KqRA+K3y/NL/oYRqQvpNJYvLm+LY1U59Q=
-github.com/aws/aws-sdk-go v1.42.52/go.mod h1:OGr6lGMAKGlG9CVrYnWYDKIyb829c6EVBRjxqjmPepc=
-github.com/aws/aws-sdk-go v1.44.34 h1:+ZtWIbtSGLNB99P8eBrxXfJZgiIouzUbpkf/MNxq2yQ=
-github.com/aws/aws-sdk-go v1.44.34/go.mod h1:y4AeaBuwd2Lk+GepC1E9v0qOiTws0MIWAX4oIKwKHZo=
 github.com/aws/aws-sdk-go-v2 v1.16.3/go.mod h1:ytwTPBG6fXTZLxxeeCCWj2/EMYp/xDUgX+OET6TLNNU=
```
Comment (AdamTylerLynch) [inline go.sum:27]: Please remove the go.sum changes.

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

```console
% make testacc TESTARGS='-run=TestAccSiteVPNConnection_\|TestAccSiteVPNCustomerGateway_' PKG=ec2 ACCTEST_PARALLELISM=3
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/ec2/... -v -count 1 -parallel 3  -run=TestAccSiteVPNConnection_\|TestAccSiteVPNCustomerGateway_ -timeout 180m
=== RUN   TestAccSiteVPNConnection_basic
=== PAUSE TestAccSiteVPNConnection_basic
=== RUN   TestAccSiteVPNConnection_transitGatewayID
=== PAUSE TestAccSiteVPNConnection_transitGatewayID
=== RUN   TestAccSiteVPNConnection_tunnel1InsideCIDR
=== PAUSE TestAccSiteVPNConnection_tunnel1InsideCIDR
=== RUN   TestAccSiteVPNConnection_tunnel1InsideIPv6CIDR
=== PAUSE TestAccSiteVPNConnection_tunnel1InsideIPv6CIDR
=== RUN   TestAccSiteVPNConnection_tunnel1PreSharedKey
=== PAUSE TestAccSiteVPNConnection_tunnel1PreSharedKey
=== RUN   TestAccSiteVPNConnection_tunnelOptions
=== PAUSE TestAccSiteVPNConnection_tunnelOptions
=== RUN   TestAccSiteVPNConnection_tunnelOptionsLesser
=== PAUSE TestAccSiteVPNConnection_tunnelOptionsLesser
=== RUN   TestAccSiteVPNConnection_staticRoutes
=== PAUSE TestAccSiteVPNConnection_staticRoutes
=== RUN   TestAccSiteVPNConnection_outsideAddressTypePrivate
=== PAUSE TestAccSiteVPNConnection_outsideAddressTypePrivate
=== RUN   TestAccSiteVPNConnection_outsideAddressTypePublic
=== PAUSE TestAccSiteVPNConnection_outsideAddressTypePublic
=== RUN   TestAccSiteVPNConnection_enableAcceleration
=== PAUSE TestAccSiteVPNConnection_enableAcceleration
=== RUN   TestAccSiteVPNConnection_ipv6
=== PAUSE TestAccSiteVPNConnection_ipv6
=== RUN   TestAccSiteVPNConnection_tags
=== PAUSE TestAccSiteVPNConnection_tags
=== RUN   TestAccSiteVPNConnection_specifyIPv4
=== PAUSE TestAccSiteVPNConnection_specifyIPv4
=== RUN   TestAccSiteVPNConnection_specifyIPv6
=== PAUSE TestAccSiteVPNConnection_specifyIPv6
=== RUN   TestAccSiteVPNConnection_disappears
=== PAUSE TestAccSiteVPNConnection_disappears
=== RUN   TestAccSiteVPNConnection_updateCustomerGatewayID
=== PAUSE TestAccSiteVPNConnection_updateCustomerGatewayID
=== RUN   TestAccSiteVPNConnection_updateVPNGatewayID
=== PAUSE TestAccSiteVPNConnection_updateVPNGatewayID
=== RUN   TestAccSiteVPNConnection_updateTransitGatewayID
=== PAUSE TestAccSiteVPNConnection_updateTransitGatewayID
=== RUN   TestAccSiteVPNConnection_vpnGatewayIDToTransitGatewayID
=== PAUSE TestAccSiteVPNConnection_vpnGatewayIDToTransitGatewayID
=== RUN   TestAccSiteVPNConnection_transitGatewayIDToVPNGatewayID
=== PAUSE TestAccSiteVPNConnection_transitGatewayIDToVPNGatewayID
=== RUN   TestAccSiteVPNCustomerGateway_basic
=== PAUSE TestAccSiteVPNCustomerGateway_basic
=== RUN   TestAccSiteVPNCustomerGateway_disappears
=== PAUSE TestAccSiteVPNCustomerGateway_disappears
=== RUN   TestAccSiteVPNCustomerGateway_privateIpv4
=== PAUSE TestAccSiteVPNCustomerGateway_privateIpv4
=== RUN   TestAccSiteVPNCustomerGateway_tags
=== PAUSE TestAccSiteVPNCustomerGateway_tags
=== RUN   TestAccSiteVPNCustomerGateway_deviceName
=== PAUSE TestAccSiteVPNCustomerGateway_deviceName
=== RUN   TestAccSiteVPNCustomerGateway_4ByteASN
=== PAUSE TestAccSiteVPNCustomerGateway_4ByteASN
=== RUN   TestAccSiteVPNCustomerGateway_certificate
=== PAUSE TestAccSiteVPNCustomerGateway_certificate
=== CONT  TestAccSiteVPNConnection_basic
=== CONT  TestAccSiteVPNConnection_specifyIPv6
=== CONT  TestAccSiteVPNCustomerGateway_basic
--- PASS: TestAccSiteVPNCustomerGateway_basic (27.10s)
=== CONT  TestAccSiteVPNCustomerGateway_certificate
--- PASS: TestAccSiteVPNCustomerGateway_certificate (81.61s)
=== CONT  TestAccSiteVPNCustomerGateway_4ByteASN
--- PASS: TestAccSiteVPNCustomerGateway_4ByteASN (26.08s)
=== CONT  TestAccSiteVPNCustomerGateway_deviceName
--- PASS: TestAccSiteVPNCustomerGateway_deviceName (25.97s)
=== CONT  TestAccSiteVPNCustomerGateway_tags
--- PASS: TestAccSiteVPNConnection_basic (207.41s)
=== CONT  TestAccSiteVPNCustomerGateway_privateIpv4
--- PASS: TestAccSiteVPNCustomerGateway_tags (57.40s)
=== CONT  TestAccSiteVPNCustomerGateway_disappears
--- PASS: TestAccSiteVPNCustomerGateway_privateIpv4 (26.33s)
=== CONT  TestAccSiteVPNConnection_updateCustomerGatewayID
--- PASS: TestAccSiteVPNCustomerGateway_disappears (21.61s)
=== CONT  TestAccSiteVPNConnection_updateVPNGatewayID
--- PASS: TestAccSiteVPNConnection_specifyIPv6 (415.75s)
=== CONT  TestAccSiteVPNConnection_disappears
--- PASS: TestAccSiteVPNConnection_disappears (197.08s)
=== CONT  TestAccSiteVPNConnection_staticRoutes
--- PASS: TestAccSiteVPNConnection_updateCustomerGatewayID (468.45s)
=== CONT  TestAccSiteVPNConnection_specifyIPv4
--- PASS: TestAccSiteVPNConnection_updateVPNGatewayID (491.30s)
=== CONT  TestAccSiteVPNConnection_tags
--- PASS: TestAccSiteVPNConnection_staticRoutes (204.16s)
=== CONT  TestAccSiteVPNConnection_ipv6
--- PASS: TestAccSiteVPNConnection_tags (408.47s)
=== CONT  TestAccSiteVPNConnection_enableAcceleration
--- PASS: TestAccSiteVPNConnection_ipv6 (490.09s)
=== CONT  TestAccSiteVPNConnection_outsideAddressTypePublic
--- PASS: TestAccSiteVPNConnection_specifyIPv4 (613.60s)
=== CONT  TestAccSiteVPNConnection_updateTransitGatewayID
--- PASS: TestAccSiteVPNConnection_updateTransitGatewayID (1619.19s)
=== CONT  TestAccSiteVPNConnection_transitGatewayIDToVPNGatewayID
--- PASS: TestAccSiteVPNConnection_outsideAddressTypePublic (1873.90s)
=== CONT  TestAccSiteVPNConnection_vpnGatewayIDToTransitGatewayID
--- PASS: TestAccSiteVPNConnection_enableAcceleration (2104.59s)
=== CONT  TestAccSiteVPNConnection_outsideAddressTypePrivate
--- PASS: TestAccSiteVPNConnection_transitGatewayIDToVPNGatewayID (637.06s)
=== CONT  TestAccSiteVPNConnection_tunnel1InsideCIDR
--- PASS: TestAccSiteVPNConnection_tunnel1InsideCIDR (199.07s)
=== CONT  TestAccSiteVPNConnection_tunnel1InsideIPv6CIDR
--- PASS: TestAccSiteVPNConnection_vpnGatewayIDToTransitGatewayID (736.98s)
=== CONT  TestAccSiteVPNConnection_tunnelOptionsLesser
--- PASS: TestAccSiteVPNConnection_tunnel1InsideIPv6CIDR (403.67s)
=== CONT  TestAccSiteVPNConnection_tunnelOptions
--- PASS: TestAccSiteVPNConnection_tunnelOptions (203.89s)
=== CONT  TestAccSiteVPNConnection_transitGatewayID
--- PASS: TestAccSiteVPNConnection_outsideAddressTypePrivate (1347.85s)
=== CONT  TestAccSiteVPNConnection_tunnel1PreSharedKey
--- PASS: TestAccSiteVPNConnection_tunnel1PreSharedKey (202.65s)
--- PASS: TestAccSiteVPNConnection_transitGatewayID (426.92s)
--- PASS: TestAccSiteVPNConnection_tunnelOptionsLesser (1248.04s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/ec2	5760.331s
```

**Comment by ewbankkit** [comment]
```console
% make providerlint
==> Checking source code with providerlint...
```

**Comment by ewbankkit** [comment]
@NetDevAutomate Thanks for the contribution :tada: :clap:.

**Pull request merged**